File: python/materials.py
# coding: utf-8
# error_codes.py

from __future__ import annotations

from Map import Map
from pedigree import PedigreeTable


#################### Materials ####################

class Materials:
	def __init__(self, path_m: str, m: Map, path_p: str, p: PedigreeTable):
		self.path_map: str = path_m
		self.geno_map: Map = m
		self.chr_maps: list[Map] = self.geno_map.create_chr_maps()
		
		self.path_ped: str = path_p
		self.ped: PedigreeTable = p
	
	def iter_chr_maps(self) -> Iterator[Map]:
		if self.chr_maps:
			return iter(self.chr_maps)
		else:
			while True:
				yield Map.default_map()
	
	def total_cM(self) -> float:
		return sum(m.total_cM() for m in self.chr_maps)
	
	def display_map_info(self):
		s = "Genetic Map : "
		if self.geno_map.is_empty():
			s += "default map(1Mbp=1cM)."
		else:
			s += self.path_map + '\n'
			s += "%d chrmosomes %f cM." % (len(self.chr_maps), self.total_cM())
		print(s)
	
	@staticmethod
	def create(path_map: str, path_ped: str) -> Materials:
		geno_map = Map.read(path_map)
		ped = PedigreeTable.read(path_ped)
		return Materials(path_map, geno_map, path_ped, ped)

File: src/materials.cpp
#include <iostream>

#include "../include/materials.h"
#include "../include/Map.h"
#include "../include/Pedigree.h"
#include "../include/common.h"

using namespace std;


//////////////////// Materials ////////////////////

Materials::Materials(const string& path_m, const Map *m,
						const string& path_p, const PedigreeTable *p) :
									path_map(path_m), geno_map(m),
									chr_maps(Map::create_chr_maps(m)),
									path_ped(path_p), ped(p) { }

Materials::~Materials() {
	delete geno_map;
	Common::delete_all(chr_maps);
	delete ped;
}

const Map *Materials::get_chr_map(int i) const {
	if(geno_map->is_empty())
		return chr_maps[0];
	else
		return chr_maps[i];
}

double Materials::total_cM() const {
	double	length = 0.0;
	for(auto p = chr_maps.begin(); p != chr_maps.end(); ++p)
		length += (*p)->total_cM();
	return length;
}

void Materials::display_map_info() const {
	cout << "Genetic Map : ";
	if(geno_map->is_empty()) {
		cout << "default map(1Mbp=1cM)." << endl;
	}
	else {
		cout << path_map << endl;
		cout << chr_maps.size() << " chrmosomes "
								<< total_cM() << " cM." << endl;
	}
}

Materials *Materials::create(const string& path_map, const string& path_ped) {
	const auto	*geno_map = Map::read(path_map);
	try {
		const auto	*ped = PedigreeTable::read(path_ped);
		return new Materials(path_map, geno_map, path_ped, ped);
	}
	catch(const std::exception& e) {
		delete geno_map;
		throw;
	}
}
